File: frontend/src/utils/dilutions.ts
/**
 * Canonical dilution representation.
 * Always stored as the denominator N in "1:N".
 * Display as "1:N".
 * For calculations, the dilution factor = 1/N.
 */
export interface ParsedDilution {
  /** The denominator N in 1:N */
  denominator: number
  /** Original text before parsing (preserved for display/audit) */
  raw: string
  /** Whether parsing was confident */
  confident: boolean
}

/**
 * Parse messy dilution text into structured form.
 * Supports formats:
 *   "1:100", "1/100", "1 to 100"  -> { denominator: 100 }
 *   "100"                          -> { denominator: 100 }
 *   "1:50-1:100"                   -> { denominator: 50 } (takes the more concentrated)
 *   "1:100 (flow)"                 -> { denominator: 100 } (strips parenthetical notes)
 *   ""  or unparseable             -> null
 */
export function parseDilution(text: string | null | undefined): ParsedDilution | null {
  if (!text || !text.trim()) return null
  const raw = text.trim()

  // Strip parenthetical notes: "1:100 (flow)" -> "1:100"
  const cleaned = raw.replace(/\s*\(.*?\)\s*/g, '').trim()

  // Range format: "1:50-1:100" or "1:50 - 1:100" -> take the lower N (more concentrated)
  const rangeMatch = cleaned.match(/1\s*[:/]\s*(\d+)\s*[-\u2013]\s*1\s*[:/]\s*(\d+)/)
  if (rangeMatch) {
    const a = parseInt(rangeMatch[1])
    const b = parseInt(rangeMatch[2])
    return { denominator: Math.min(a, b), raw, confident: true }
  }

  // Standard formats: "1:N", "1/N"
  const stdMatch = cleaned.match(/1\s*[:/]\s*(\d+)/)
  if (stdMatch) {
    return { denominator: parseInt(stdMatch[1]), raw, confident: true }
  }

  // "1 to N"
  const toMatch = cleaned.match(/1\s+to\s+(\d+)/i)
  if (toMatch) {
    return { denominator: parseInt(toMatch[1]), raw, confident: true }
  }

  // Bare number: "100" -> assume 1:100
  const bareMatch = cleaned.match(/^(\d+)$/)
  if (bareMatch) {
    return { denominator: parseInt(bareMatch[1]), raw, confident: false }
  }

  return null
}

/**
 * Format a denominator for display as "1:N".
 */
export function formatDilution(denominator: number | null | undefined): string {
  if (denominator == null) return ''
  return '1:' + denominator
}

/**
 * Format for input field: just show the number.
 */
export function dilutionToInputValue(denominator: number | null | undefined): string {
  if (denominator == null) return ''
  return String(denominator)
}

File: backend/services/dilutions.py
from __future__ import annotations

import re


def parse_dilution(text: str | None) -> int | None:
    """Parse messy dilution text into the denominator N of 1:N.

    Supports: "1:100", "1/200", "1 to 100", "100", "1:50-1:100" (takes lower N),
    "1:100 (flow)" (strips parenthetical notes). Returns None if unparseable.
    """
    if not text or not text.strip():
        return None
    raw = text.strip()
    # Strip parenthetical notes
    cleaned = re.sub(r"\s*\(.*?\)\s*", "", raw).strip()

    # Range format: "1:50-1:100" or "1:50 - 1:100"
    range_match = re.search(r"1\s*[:/]\s*(\d+)\s*[-\u2013]\s*1\s*[:/]\s*(\d+)", cleaned)
    if range_match:
        a = int(range_match.group(1))
        b = int(range_match.group(2))
        return min(a, b)

    # Standard: "1:N", "1/N"
    std_match = re.search(r"1\s*[:/]\s*(\d+)", cleaned)
    if std_match:
        return int(std_match.group(1))

    # "1 to N"
    to_match = re.search(r"1\s+to\s+(\d+)", cleaned, re.IGNORECASE)
    if to_match:
        return int(to_match.group(1))

    # Bare number: "100"
    bare_match = re.fullmatch(r"(\d+)", cleaned)
    if bare_match:
        return int(bare_match.group(1))

    return None
